pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,252 @@
+parent schemas. The subschemas that are not part of its lexical scope have been
+grayed out:
+
+![The lexical scope of a JSON Schema represented as a tree structure](/img/posts/2024/understanding-lexical-dynamic-scopes/ancestor.webp)
+
+In comparison to the dynamic scope, the lexical scope of any subschema can be
+statically determined without taking
+[instances](https://json-schema.org/learn/glossary#instance) into account, just
+as we did here.
+
+### The Schema Resource Boundary
+
+The `$id` keyword defines the URI of a schema. While this keyword is typically
+set at the top level, any subschema may declare it to distinguish itself with a
+different URI. For example, the following schema defines 4 identifiers, some
+relative and some absolute:
+
+![A JSON Schema with multiple identifiers](/img/posts/2024/understanding-lexical-dynamic-scopes/multiple-ids.webp)
+
+[In JSON Schema
+parlance](https://json-schema.org/draft/2020-12/json-schema-core#name-root-schema-and-subschemas-),
+we say that the `$id` keyword introduces a new *schema resource*, and that the
+top level schema resource is referred to as the *root schema resource*.
+
+Consider the following example of schema resources. This schema consists of 3
+schema resources, each highlighted using a different color: the root schema
+resource (red), the schema resource at `/properties/foo` (blue), and the schema
+resource at `/properties/bar` (green). Note that the subschema at
+`/properties/baz` is part of the root schema resource:
+
+![A JSON Schema that consists of 3 schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/schema-resources.webp)
+
+Going back to scopes, <i>an important rule to remember is that the lexical
+scope of a schema does not extend past its schema resource</i>.
+
+As an example of the relationship between schema resources and lexical scopes,
+consider the `$anchor` keyword that defines a [location-independent
+identifier](https://json-schema.org/draft/2020-12/json-schema-core#section-8.2.2)
+for a schema. While declaring the same anchor identifier more than once in the
+same schema resource is an error, it is possible to declare the same anchor
+identifier on different schema resources within the same schema. This is
+possible because *schema resources do not share their lexical scopes*:
+
+![Defining multiple anchors with the same name on different schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/static-anchors.webp)
+
+### Changing Scopes With References
+
+Now that we understand lexical scopes and their boundaries within schema
+resources, lets consider the effect that following references has on them.
+<i>When evaluation encounters a reference keyword, it abandons the lexical
+scope of the reference schema and enters the lexical scope of the destination
+schema</i>. In other words, the evaluation process needs to re-calculate the
+lexical scope of the schema it ends up at after following the reference.
```

Comment:
> The subschema you end up in is simply the beginning of the lexical scope for that document

This is the part I guess I'm confused about. I thought the lexical scope consists of the parents of the schema, and that's why i.e. if you jump to a subschema of a schema resource that defines its own `$schema` you inherit that new dialect (because `$schema` was on the lexical scope of the target).

Reply:
This is the part of the spec I'm referring to:

> Note that some keywords, such as "$schema", apply to the lexical scope of the entire schema resource, and therefore MUST only appear in a schema resource's root schema.
> https://json-schema.org/draft/2020-12/json-schema-core#section-7.1-4

That seems to imply that `$schema` applies to a lexical scope because its one of the structural parents of the schema, but maybe I'm getting it wrong

Reply:
Also, on why I used the term "calculate" here:

> While following the reference keyword moves processing from one lexical scope into a different one
> https://json-schema.org/draft/2020-12/json-schema-core#section-7.1-6

The specification seems to say we "move" from one lexical scope to another one when following any reference (not just dynamic), and to me "move" means "recalculate what the other one is" instead of just "discard the one you have until that point"

There is also this part:

> A keyword on the far side of that reference that resolves information through the dynamic scope will consider the originating side of the reference to be their dynamic parent, rather than examining the local lexically enclosing parent.

Which contrasts how the dynamic scope is more about the evaluation path while the lexical scope is more about its "enclosing parent"

Reply:
It may help to understand that "lexically" just means "as written".  So here, "local lexically enclosing parent" doesn't imply that the parent is part of the scope, but rather that a dynamic ref doesn't consider the parent "as written" like a static ref might.

It is rather confusing to make that clarification if the lexical scope doesn't include the local parent, though.

Reply:
I mentioned in another comment that some keywords have a quality of applying to the whole document. These include `$id`, `$schema`, `$anchor`, and `$dynamicAchor`. Whatever subschema you're in, those keywords are in scope. So, when you enter `https://my-dialect.com#/items`, the `$id` and `$schema` are part of the evaluation of that schema even tho they are in a different sub-schema.

It's not as simple as saying the parents schemas are in the same scope, because it's very possible that there are schemas in between the root with `$id` and `$schema` and the schema you're evaluating that are never part of the scope.

Reply:
@jdesrosiers I think you're mis-reading the schema Juan posted.  `https://my-dialect.com` is a new `$schema`, not the `$id` of the subschema.

Reply:
> Whatever subschema you're in, those keywords are in scope

That sounds to me like yet a 3rd type of scoping. Risking making things more complicated, maybe we should give this a name?